test/passing/gen/gen.ml

```diff
@@ -93,6 +97,49 @@ let cmd should_fail args =
        (run %s))|} cmd_string
   else spf {|(run %s)|} cmd_string
 
+let one_styling_test ~extra_deps ~enabled_if_line ~test_name ~base_test_name
+    ~should_fail ~opts ~output_name ~extra_suffix =
+  Printf.sprintf
+    {|
+(rule
+ (deps tests/.ocamlformat %s)%s
+ (package ocamlformat)
+ (action
+  (with-stdout-to %s.%sstdout
+   (with-stderr-to %s.%sstderr
+     %s))))
+
+(rule
+ (alias runtest)%s
+ (package ocamlformat)
+ (action (diff %s %s.%sstdout)))
+
+(rule
+ (alias runtest)%s
+ (package ocamlformat)
+ (action (diff tests/%s.%serr %s.%sstderr)))
+|}
+    extra_deps enabled_if_line test_name extra_suffix test_name extra_suffix
+    (cmd should_fail (["%{bin:ocamlformat}"] @ opts @ [dep base_test_name]))
+    enabled_if_line output_name test_name extra_suffix enabled_if_line
+    test_name extra_suffix test_name extra_suffix
+
+let one_js_coverage_test ~test_name ~should_test =
+  let test_sigil, (err_msg : _ format) =
+    if should_test then ("", "%s has both a [.js-ref] and a [.why-no-js]!")
+    else (" !", "%s has neither a [.js-ref], nor a [.why-no-js]!")
+  in
+  let err_msg = Printf.sprintf err_msg test_name in
+  Printf.sprintf
+    {|
+(rule
+ (alias runtest)
+ (enabled_if (<> %%{os_type} Win32))
+ (package ocamlformat)
+ (action (system "if [%s -f tests/%s.why-no-js ]; then echo '%s'; exit 1; fi")))
+|}
```

Comment:
We know during the running of `gen.ml` whether or not both/neither are present; lets only print out a rule like this if we know it's going to fail.